### ISSUE TITLE
Pooled IPs can get stale on redis cluster instance rescheduling

### ISSUE THREAD

**Issue by roma-glushko** [alisaifee/coredis#149]
While using `coredis.RedisCluster` against a self-hosted redis cluster chart (like [Bitnami Redis Cluster chart](https://github.com/bitnami/charts/tree/main/bitnami/redis-cluster) that deploys redis as a statefulset with a regular service resource that includes all redis nodes), I have noticed that at some point RedisCluster starts to try to access IPs that are not part of the cluster anymore. I gather they were a part of the cluster and those IPs were occupied by Redis cluster instances, but then perhaps Kubernetes decided to reschedule of of them what gave them another IPs. 

If you get unfortunate enough, those IPs might be used in one of your redis client instance what in turn cause connection errors that could be fixed by restarting corresponding pods (what effectively refresh IPs in the force way). 

The `RedisCluster` client is initialized in the following way:

```python
cluster_client = RedisCluster(
        host="redis",  # which refers to the service (a ClusterIP type) resource that points to all statefulset's pods
        port=6379,
        username=None,
        password="SOMEPASSWORD",
        ssl=True, # It was tested with False as well resulting in the same issue
        ssl_cert_reqs="none",
        skip_full_coverage_check=True,
        read_from_replicas=False,
    )
```

Do you folks have any idea what's going wrong here?

**Comment by alisaifee** [comment]
At the moment the cluster client will perform a refresh of the cluster nodes when either a clusterdown or movederror is encountered ([reference](https://github.com/alisaifee/coredis/blob/master/coredis/client/cluster.py#L751)). 

Is my understanding correct that in your use case this is not the case and the ip of a previous node is simply pointing to "something else" and the exception is a ConnectionError which simply gets retried with the same set of nodes as before?

**Comment by alisaifee** [comment]
This is unfortunately the category of error handling that can't easily be tested with automated test. This is probably asking a lot but is there any chance you could provide a set of scripts or instructions that would let me reproduce the cluster state?

**Comment by roma-glushko** [comment]
> This is probably asking a lot but is there any chance you could provide a set of scripts or instructions that would let me reproduce the cluster state?

@alisaifee let me try to provide something useful for you. Would an example in tilt/k3d setup work for you? I will try to extract relevant pieces and pack it in a way you could reproduce locally. Maybe that will help to hunt the problem down by itself.

**Comment by alisaifee** [comment]
tilt + k3d works for me! I think for the "service" portion a forever looping subscriber on an arbitrary topic will probably be enough?

**Comment by alisaifee** [comment]
Actually, a second service that is forever publishing as well would be useful.. since the publishing aspect of the code goes through the regular redis/rediscluster client `execute_command` flow whereas the subscriber uses it's own connection handling.

**Comment by roma-glushko** [comment]
Hey @alisaifee 👋 
Getting back to you with [a sample application](https://github.com/roma-glushko/issue-reports/tree/main/coredis/issue149) that consists of three main components:
- a Redis Cluster with 6 nodes
- a `heartbeat` component that publishes messages periodically 
- a `tracker` component that listens to messages and print them to stdout

<img width="1920" alt="Screenshot 2023-03-01 at 19 49 59" src="https://user-images.githubusercontent.com/9402690/222233501-f627a108-7708-4c02-995b-d74f76f89b13.png">

If you try to restart each redis instance as I was mentioning, you will catch a situation when both heartbeat and tracker hung up for a while and then start erroring. At the same time, the redis cluster will recover from the restart super fast, but unfortunately, the components will still failing with no signs of self-healing.

**Comment by alisaifee** [comment]
Thank you @roma-glushko. These are probably the best reproduction steps / artifact that I've ever come across for a bug report. Great work!

I was able to get it up and running and also reproduce the error both in the `heartbeat` & `tracker` services. I used the `Kill` action in `k9s` sequentially on all the redis instances to reproduce the error.

I can clearly see why it's happening and as I had guessed there are two different buggy code paths that result in the stale connections.
- publishing: ConnectionErrors are caught [here](https://github.com/alisaifee/coredis/blob/master/coredis/client/cluster.py#L746) in the cluster client and this condition doesn't force a cluster topology re-initialization but it does trigger a retry with a `random_node`. This may or may not work since it's ... random. Since the command is `PUBLISH` redis cluster will accept it on any node since that's not a sharded operation and not result in a `MOVED` error which would have otherwise resulted in a cluster re-initialization [here](https://github.com/alisaifee/coredis/blob/master/coredis/client/cluster.py#L757). One possible fix here is that if a connection error occurs for a non sharded operation we can try to influence triggering a re-initialization.
- subscribing: The subscriber pins the connection to a node based on the hash of the topic at the point of the first subscription. After this if a call to `get_message` results in a ConnectionError - we simply disconnect the connection and reconnect (which is still pinned to the stale IP) and retry once [here](https://github.com/alisaifee/coredis/blob/master/coredis/commands/pubsub.py#L208) which is not useful. I think a possible solution here could be to fetch a new connection from the pool instead of trying to reuse the original connection and while doing so consider re-initializing the cluster.

The part which I'm trying to figure out right now (and need help with) is - how can I distinguish a temporary connection error to an otherwise healthy node identified by an IP `x` from an IP that no longer points to a redis instance.

**Comment by roma-glushko** [comment]
Hey @alisaifee appreciate you found a chance to give it a look 🙏 

> These are probably the best reproduction steps / artifact that I've ever come across for a bug report. Great work!

Haha, nice! I think we need more of this in the open source community 🙌 

> how can I distinguish a temporary connection error to an otherwise healthy node identified by an IP x from an IP that no longer points to a redis instance.

That's a good question. Let's just think out loud and assume the following for a moment:

* A1: Transient network errors happen but they are rare. So it's unlikely to get a failed request on retry (because it is super unlikely that two rare network errors happen two times in a row)
* A2: Redis instance restarts/failures are roughly unconnected. In reality this is a bit off from what I have seen. For example, if cluster autoscaling is the reason for instances to get restarted, then usually a few instances get restarted in at the same time (this can be controlled by [PodDisruptionBugets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) but still). Also, I'm not sure what happens to the cluster when the current primary instance goes down. Would a new master be elected? My empirical experience shows that the cluster (the same 6 nodes cluster like in the sample app) is extremely fragile when the master is down. So this assumption trips on these two cases, but generally it should be fairly reasonable thing to assume to simplify the model.
* A3: We have a list of all redis instances that was actual at least at some point of time
* A4: `ConnectionError` happens either because of transient network errors, or IP:port doesn't point to a redis instance anymore, or a Redis instance is in a process of starting, but not yet ready to serve the request (e.g. ConnectionError doesn't happen when an instance is overloaded (?))

Then, we could think of at least two straightforward ways to answer the question:

* Pick up an instance A from the pool, send it a request and it fails with `ConnectionError`, retry (possibly in exponential manner) the request and get another `ConnectionError`. At that point, it would be clear that something fishy is going on with the instance A. It's either moved to another IP:port address in the cluster or happen to be initializing still (A4) (A1 is unlikely). Doing another retry with reasonable delay could cross the second option out leaving us with stale topology assumption as the most probable one.
* Since we have information on all instances in the cluster (A3), we could pull an instance A from the pool, do a request and fail with `ConnectionError`. We record that fail and draw the next instance (A2) at random keeping the failed one aside. Get another `ConnectionError` and record that fail as well. Repeat the same process against remaining/untested instances until we bump into some configurable threshold. Generally speaking, the threshold will depend on the cluster size and some redis cluster specifics (Redis cluster should have at least three nodes to meet a quorum).

Redis-py has even "tougher" logic on `ConnectionError` (https://github.com/redis/redis-py/blob/master/redis/cluster.py#L1118). It seems to reinitialize everything on the first `ConnectionError`/`TimeoutError` occurrence. Depending on how expensive it is to do reinitialization, that maybe a fairly good option.

**Comment by alisaifee** [comment]
Thanks for brainstorming the potential causes of `ConnectionError` due to node instability in the redis cluster. I'm trying to work out a slightly less brute force approach than to just trigger a reinitialization on any `ConnectionError` in the cluster with some of those ideas in mind.

Having said that I do have a very rough implementation that does in fact use a brute force approach to try and handle node restarts in the `cluster-node-restart-handling` branch. I've tested it out with the sample you provided and so far it appears that both `heartbeat` and `tracker` services survive the restarts. One minor change you have to make to your code though is to set `nodemanager_follow_cluster` to `False` since the default `True` results in the reinitializing step using the ips reported by the cluster in the first initialization - which does not include the `redis:6379` node.

Below are the patches you can try:

```patch
diff --git a/coredis/issue149/pyproject.toml b/coredis/issue149/pyproject.toml
index 5d8bca5..c7ff691 100644
--- a/coredis/issue149/pyproject.toml
+++ b/coredis/issue149/pyproject.toml
@@ -13,7 +13,8 @@ packages = [
 [tool.poetry.dependencies]
 python = "^3.9"
 pydantic = "^1.10.2"
-coredis = "^4.10.2"
+#coredis = "^4.10.2"
+coredis = { git = "https://github.com/alisaifee/coredis.git", branch = "cluster-node-restart-handling" }
 
 [tool.poetry.dev-dependencies]
 ruff = "^0.0.152"

```

```patch
diff --git a/coredis/issue149/common/broker.py b/coredis/issue149/common/broker.py
index 453bda9..35a7768 100644
--- a/coredis/issue149/common/broker.py
+++ b/coredis/issue149/common/broker.py
@@ -24,6 +24,7 @@ class BrokerConfig(BaseSettings):
         default=False, env="BROKER_REQUIRE_FULL_COVERAGE"
     )
     read_from_replicas: bool = Field(default=False, env="BROKER_READ_FROM_REPLICAS")
+    follow_cluster: bool = Field(default=True, env="BROKER_FOLLOW_CLUSTER")
 
 
 def create_redis_cluster_client(broker_config: BrokerConfig) -> RedisCluster:
@@ -38,8 +39,8 @@ def create_redis_cluster_client(broker_config: BrokerConfig) -> RedisCluster:
         password=password.get_secret_value() if password else None,
         skip_full_coverage_check=not broker_config.cluster_require_full_coverage,
         read_from_replicas=broker_config.read_from_replicas,
+        nodemanager_follow_cluster=broker_config.follow_cluster,
     )
-
     return cluster_client
 
 

```

**Comment by roma-glushko** [comment]
Hey @alisaifee 👋 
Thank you for the blazing fast PR with improvements ⚡ ⚡ ⚡ 

I have applied [all changes](https://github.com/roma-glushko/issue-reports/pull/2) that you have suggested and I do see positive differences after that. I could not get clients broken temporarily or permanently like it had been before which is great 👍 

At the same time, now we have got a notion of recovery time. I feel like that time is a bit too long than I would ideally expect. 

For example, on this screencast, we could see that redis cluster itself got recovered fairly quickly, while it took a way longer for the heartbeat to back on the level and begin to publish new messages:

https://user-images.githubusercontent.com/9402690/223123848-531e183a-3ad5-439a-be2b-64ebce983b9d.mov

That may be related to the following line where we resetting cluster topology only after 9 failed retries with the current one:

- https://github.com/alisaifee/coredis/pull/151/files#diff-9720504ad266246d98b42895db5627b3bd195d06f2a75398cee1355fce37bd72R750

Other than that, I was able to miss a few events between 68 and 80 (however, heartbeat sent them):

<img width="1920" alt="Screenshot 2023-03-06 at 15 14 33" src="https://user-images.githubusercontent.com/9402690/223126019-0c5eafd0-cfe5-424d-82f5-c6680bedc25c.png">
<img width="1127" alt="Screenshot 2023-03-06 at 15 14 45" src="https://user-images.githubusercontent.com/9402690/223126081-c065e30f-7193-4a0f-9bc6-198c69ad2119.png">

I think I might have restarted an instance that the tracker was listening to, while heartbeat was unaffected. Not sure we can much about that (especially in the current setup with only one tracker). Maybe optimizing the recovery time on both pub/sub sides could help to mitigate such situation.

**Comment by alisaifee** [comment]
Thank you for testing out the WIP and glad to hear that at least there is some improvement!

I too was taken aback by the recovery time - more so because the delay between retries doesn't sum up to some of the worst case recovery times I encountered. There are some minor improvements in the `cluster-node-restart-handling` branch including retries for sharded subscribers (Since you're using redis cluster with pubsub this is something you could consider for improved throughput. Some documentation [here](https://coredis.readthedocs.io/en/latest/handbook/pubsub.html#sharded-pub-sub)). The code in that branch used with the configuration you shared still results in an unacceptable recovery time (IMO) - however I believe that is caused by the fact the default `connect_timeout` is unset and therefore in certain scenarios a retry results in the cluster topology not having been updated and an attempt to connect to a retiring node holding up the recovery. Passing `connect_timeout=1` for example when initializing the client in my tests shows an almost immediate recovery for both the publisher and the subscriber.

**Comment by alisaifee** [comment]
I've merged some of the improvements into master. These include:
- Retry policy applied to both cluster & sharded cluster subscribers. The retry policy is also exposed so that you can control the parameters ([some initial documentation here](https://coredis.readthedocs.io/en/latest/api.html#retries))
- publish command now uses a random node and on connection error results in a cluster topology refresh
- timeout to establish a connection now raises a `ConnectionError` which will trigger the appropriate retries

**Comment by roma-glushko** [comment]
Hey @alisaifee 👋 sorry for delay, signed off for a few prev days.

> There are some minor improvements in the cluster-node-restart-handling branch including retries for sharded subscribers (Since you're using redis cluster with pubsub this is something you could consider for improved throughput. Some documentation [here](https://coredis.readthedocs.io/en/latest/handbook/pubsub.html#sharded-pub-sub))

Yes, I would love to use shared clustering eventually. As of now, we have some part of the codebase that doesn't compatible with the sharded mode. It's related to another subscriber based on SocketIO that listens to emitted messages but doesn't seem to support sharded cluster subscriptions for the time being 🤷 

> however I believe that is caused by the fact the default connect_timeout is unset and therefore in certain scenarios a retry results in the cluster topology not having been updated and an attempt to connect to a retiring node holding up the recovery. Passing connect_timeout=1 for example when initializing the client in my tests shows an almost immediate recovery for both the publisher and the subscriber.

Niiiiice! This is exactly what I needed 👍  I would maybe put there some limits (instead of None) as a default since it's always good to put limits/backpressure in the system.

When do you to create a new release that will include these improvements? I would like to give it a test in the real setup.

**Comment by alisaifee** [comment]
@roma-glushko no worries! Thanks for all your help in improving this part of the library! I want to take the `RetryPolicy` implementation a bit further and try to use it more generally and make it more accessible to library users but I think that can wait for a future release.

I'll iron out some rough edges and aim to release `4.11` by EOD.

Regarding the `None` `connect_timeout` I definitely agree that it should be something a bit safer... but I'll also leave that for a future release as there are some other bad defaults that I'd like to address together with better documentation as well.

**Comment by alisaifee** [comment]
`4.11.x` has been released and contains the fixes discussed above.

**Comment by alisaifee** [comment]
@roma-glushko just FYI, I've made some updates in master and a pre-release [4.12.0b1](http://127.0.0.1:18005/release_notes.html#v4-12-0b1) which exposes `retry_policy` in the `Redis` & `RedisCluster` constructors as well to provide better control over retries when sending the `PUBLISH` commands. Perhaps you want to check that out if it would be an improvement for you.
